File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/java/src/main/java/com/example/dp/UniquePaths.java
package com.example.dp;

import java.util.HashMap;
import java.util.Map;

public final class UniquePaths {
    private UniquePaths() {
    }

    public static Result memoized(int rows, int cols) {
        return new Result("Unique Paths", "memoization", solve(0, 0, rows, cols, new HashMap<>()), Map.of());
    }

    private static int solve(int row, int col, int rows, int cols, Map<String, Integer> memo) {
        if (row == rows - 1 && col == cols - 1) {
            return 1;
        }
        if (row >= rows || col >= cols) {
            return 0;
        }
        String key = row + ":" + col;
        if (memo.containsKey(key)) {
            return memo.get(key);
        }
        int value = solve(row + 1, col, rows, cols, memo) + solve(row, col + 1, rows, cols, memo);
        memo.put(key, value);
        return value;
    }

    public static Result tabulated(int rows, int cols) {
        int[][] dp = new int[rows][cols];
        for (int row = 0; row < rows; row++) {
            dp[row][0] = 1;
        }
        for (int col = 0; col < cols; col++) {
            dp[0][col] = 1;
        }
        for (int row = 1; row < rows; row++) {
            for (int col = 1; col < cols; col++) {
                dp[row][col] = dp[row - 1][col] + dp[row][col - 1];
            }
        }
        return new Result("Unique Paths", "tabulation", dp[rows - 1][cols - 1], Map.of("rows", rows, "cols", cols));
    }

    public static Result spaceOptimized(int rows, int cols) {
        int[] dp = new int[cols];
        for (int col = 0; col < cols; col++) {
            dp[col] = 1;
        }
        for (int row = 1; row < rows; row++) {
            for (int col = 1; col < cols; col++) {
                dp[col] += dp[col - 1];
            }
        }
        return new Result("Unique Paths", "space optimized", dp[cols - 1], Map.of());
    }
}

File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/python/dp_algorithms/unique_paths.py
from __future__ import annotations

from functools import lru_cache

from dp_algorithms.common import DPResult


def memoized(rows: int, cols: int) -> DPResult:
    @lru_cache(maxsize=None)
    def solve(row: int, col: int) -> int:
        if row == rows - 1 and col == cols - 1:
            return 1
        if row >= rows or col >= cols:
            return 0
        return solve(row + 1, col) + solve(row, col + 1)

    return DPResult("Unique Paths", "memoization", solve(0, 0))


def tabulated(rows: int, cols: int) -> DPResult:
    dp = [[1] * cols for _ in range(rows)]
    for row in range(1, rows):
        for col in range(1, cols):
            dp[row][col] = dp[row - 1][col] + dp[row][col - 1]
    return DPResult("Unique Paths", "tabulation", dp[-1][-1], {"last_row": dp[-1]})


def space_optimized(rows: int, cols: int) -> DPResult:
    dp = [1] * cols
    for _ in range(1, rows):
        for col in range(1, cols):
            dp[col] += dp[col - 1]
    return DPResult("Unique Paths", "space optimized", dp[-1])
